src/StoreApi/Utilities/OrderController.php

```diff
@@ -155,26 +175,210 @@ public function validate_order_before_payment( \WC_Order $order ) {
+		$shipping_address = $order->get_address( 'shipping' );
+
+		if ( $needs_shipping && ! $this->validate_allowed_country( $shipping_address['country'], (array) wc()->countries->get_shipping_countries() ) ) {
+			throw new RouteException(
+				'woocommerce_rest_invalid_address_country',
+				sprintf(
+					// Translators: %s country code.
+					__( 'Sorry, we do not ship orders to the provided country (%s)', 'woo-gutenberg-products-block' ),
+					$shipping_address['country']
+				),
+				400,
+				[
+					'allowed_countries' => array_keys( wc()->countries->get_shipping_countries() ),
+				]
+			);
+		}
+
+		if ( ! $this->validate_allowed_country( $billing_address['country'], (array) wc()->countries->get_allowed_countries() ) ) {
+			throw new RouteException(
+				'woocommerce_rest_invalid_address_country',
+				sprintf(
+					// Translators: %s country code.
+					__( 'Sorry, we do not allow orders from the provided country (%s)', 'woo-gutenberg-products-block' ),
+					$billing_address['country']
+				),
+				400,
+				[
+					'allowed_countries' => array_keys( wc()->countries->get_allowed_countries() ),
+				]
+			);
+		}
+
+		if ( $needs_shipping ) {
+			$this->validate_address_fields( $shipping_address, 'shipping', $errors );
+		}
+		$this->validate_address_fields( $billing_address, 'billing', $errors );
+
+		if ( ! $errors->has_errors() ) {
+			return;
+		}
+
+		$errors_by_code = [];
+		$error_codes    = $errors->get_error_codes();
+		foreach ( $error_codes as $code ) {
+			$errors_by_code[ $code ] = $errors->get_error_messages( $code );
+		}
+
+		// Surface errors from first code.
+		foreach ( $errors_by_code as $code => $error_messages ) {
+			throw new RouteException(
+				'woocommerce_rest_invalid_address',
+				sprintf(
+					// Translators: %s Address type.
+					__( 'There was a problem with the provided %s:', 'woo-gutenberg-products-block' ) . ' ' . implode( ', ', $error_messages ),
+					'shipping' === $code ? __( 'shipping address', 'woo-gutenberg-products-block' ) : __( 'billing address', 'woo-gutenberg-products-block' )
+				),
+				400,
+				[
+					'errors' => $errors_by_code,
+				]
+			);
+		}
+	}
+
+	/**
+	 * Check all required address fields are set and return errors if not.
 	 *
-	 * @throws \Exception Exception if invalid data is detected.
+	 * @param string $country Country code.
+	 * @param array  $allowed_countries List of valid country codes.
+	 * @return boolean True if valid.
+	 */
+	protected function validate_allowed_country( $country, array $allowed_countries ) {
+		return array_key_exists( $country, $allowed_countries );
+	}
+
+	/**
+	 * Check all required address fields are set and return errors if not.
 	 *
+	 * @param array     $address Address array.
+	 * @param string    $address_type billing or shipping address, used in error messages.
+	 * @param \WP_Error $errors Error object.
+	 */
+	protected function validate_address_fields( $address, $address_type = 'billing', \WP_Error &$errors ) {
```

Comment:
Why passing $errors by reference here? `$errors` is still a pointer to the `\WP_Error` instance so it's not creating a new copy of `\WP_Error`?